services/udp/service.go

```diff
@@ -174,45 +197,57 @@ func (s *Service) writer() {
 					logger.Database(s.config.Database), zap.Error(err))
 				atomic.AddInt64(&s.stats.BatchesTransmitFail, 1)
 			}
-
-		case <-s.done:
-			return
 		}
 	}
 }
 
-func (s *Service) serve() {
+func (s *Service) serve(ctx context.Context) {
 	defer s.wg.Done()
-
 	buf := make([]byte, MaxUDPPayload)
+
+	type readpayload struct {
+		nbytes int
+		err    error
+	}
+
+	payloadChan := make(chan readpayload)
+
+	go func() {
+		for {
+			payload := readpayload{}
+			payload.nbytes, _, payload.err = s.conn.ReadFromUDP(buf)
+			payloadChan <- payload
+		}
+	}()
+
 	for {
 		select {
-		case <-s.done:
+		case <-ctx.Done():
 			// We closed the connection, time to go.
+			s.conn.Close()
 			return
-		default:
+		case payload := <-payloadChan:
 			// Keep processing.
-			n, _, err := s.conn.ReadFromUDP(buf)
-			if err != nil {
+			if payload.err != nil {
 				atomic.AddInt64(&s.stats.ReadFail, 1)
-				s.Logger.Info("Failed to read UDP message", zap.Error(err))
+				s.Logger.Info("Failed to read UDP message", zap.Error(payload.err))
 				continue
 			}
-			atomic.AddInt64(&s.stats.BytesReceived, int64(n))
+			atomic.AddInt64(&s.stats.BytesReceived, int64(payload.nbytes))
 
-			bufCopy := make([]byte, n)
-			copy(bufCopy, buf[:n])
+			bufCopy := make([]byte, payload.nbytes)
+			copy(bufCopy, buf[:payload.nbytes])
```

Comment:
These two goroutines will race on reads / writes to the `buf`, which is read and written to concurrently.

Reply:
fixed!